Set LD_PRELOAD with buildpack script, fix paths

brian-kephart opened the pull request:
3 changes:

1. Add a line to set LD_PRELOAD with the buildpack script in /.profile.d. Setting this value in Heroku config causes errors in slug compilation because the compiler looks for the script before it's created by the buildpack. #6 
2. Update README.md to remove obsolete instructions.
3. Fix an issue that causes leading/trailing colons in path variables that had no value before being altered by the buildpack.

czj commented:
This seems to be a lot easier to use and more reliable than both previous solutions.
Thanks a lot !

brian-kephart commented:
@mojodna @nateberkopec Any chance of getting this PR reviewed?

brian-kephart commented:
I just discovered this PR is incompatible with jemalloc 5, as the filename needed for LD_PRELOAD changes in that version. There is a separate PR for the jemalloc-5.0.0 branch, which is likewise incompatible with lower versions.

nateberkopec commented:
> Fix an issue that causes leading/trailing colons in path variables that had no value before being altered by the buildpack.

@brian-kephart Can you submit this as a separate PR? This seems like a good fix.

I'm not sure we should set LD_PRELOAD automatically. That's a major change in the behavior of the buildpack. I think it's good that a) adding the buildpack by itself does not immediately change application behavior and b) one can "roll back" a jemalloc change simply by changing an env variable and c) one can turn jemalloc on for just one process or two.

For that reason, closing this for now, though I'm open to have my mind changed.

brian-kephart commented:
Thanks for looking at this.

From my perspective, the largest advantages of setting LD_PRELOAD automatically are 1) zero config required and 2) changes to jemalloc itself are handled more gracefully. For example, the name of the file required for the LD_PRELOAD config option changed from version 4 to 5. With the current implementation, upgrading to v5 requires changing your config, meaning changing the master branch to v5 requires documentation, user configuration, etc. instead of just changing a line of code.

This is in addition to the fact that using heroku config for LD_PRELOAD never worked for some people, including me (see #6).

None of which addresses your concerns, of course, just making sure the pros and cons are all accounted for.

Do you still want a separate PR for the path variables? I see you merged that into the other branch after your comment here, so I'm not sure if it's still needed.

nateberkopec commented:
Thanks for commenting @brian-kephart, I still think it's not worth it (for now, anyway).

> Do you still want a separate PR for the path variables? I see you merged that into the other branch after your comment here, so I'm not sure if it's still needed.

Nope, cherry-picked your commit.
